Pokedex/Models/PokemonResponse.swift

```diff
@@ -0,0 +1,62 @@
+//
+//  Pokemon.swift
+//  Pokedex
+//
+//  Created by krikaz on 6/9/20.
+//  Copyright © 2020 thewire. All rights reserved.
+//
+
+import Foundation
+
+private struct RawServerResponse: Decodable {
+    
+    var moves: [SingleMove]
+    var sprites: Sprite
+    var types: [SingleType]
+    var name: String
+
+    struct SingleMove: Decodable {
+        let move: MoveDetail
+    }
+    
+    struct MoveDetail: Decodable {
+        let name: String
+        let url: String
+    }
+    
+    struct Sprite: Decodable {
+        let front_default: String
+        let front_shiny: String
+    }
+
+    struct SingleType: Decodable {
+        let slot: Int
+        let type: TypeDetail
+    }
+    
+    struct TypeDetail: Decodable {
+        let name: String
+        let url: String
+    }
+}
+
+struct PokemonResponse: Decodable {
+    var name: String
+    var spriteURLString: String
+    var types: [String] = []
+    var attacks: [String] = []
+    
+    init(from decoder: Decoder) throws {
+        let rawResponse = try RawServerResponse(from: decoder)
+        
+        name = rawResponse.name
+        spriteURLString = rawResponse.sprites.front_default
+        types.append(rawResponse.types[0].type.name)
+        attacks.append(rawResponse.moves[0].move.name)
+    }
+}
```

Comment:
Can you send me a JSON of the response ? I need to see it so I can help you model it.

Reply:
https://pokeapi.co/api/v2/pokemon/alakazam
I can't attach the json file

Reply:
Here's what I would do. You use `PokemonDetailResponse` for the response of the request, and you use `PokemonDetail` in your views.

```swift
struct PokemonDetailResponse: Decodable {
    struct PokemonType: Decodable {
        struct Detail: Decodable {
            let name: String
            let url: URL
        }

        let detail: Detail
    }

    struct Move: Decodable {
        struct Detail: Decodable {
            let name: String
            let url: URL
        }

        let detail: Detail
    }

    let name: String
    let spriteURL: URL
    let types: [PokemonType]
    let moves: [Move]
}

struct PokemonDetail {
    let name: String
    let spriteURL: URL
    let types: [String]
    let attacks: [String]

    init(response: PokemonDetailResponse) {
        self.name = response.name
        self.spriteURL = response.spriteURL
        self.types = response.types.map { $0.detail.name }
        self.attacks = response.types.map { $0.detail.name }
    }
}
```

Reply:
Here's what I would do. You use `PokemonDetailResponse` for the response of the request, and you use `PokemonDetail` in your views.

```swift
struct PokemonDetailResponse: Decodable {
    struct Sprites: Decodable {
        let frontDefault: String
    }

    struct PokemonType: Decodable {
        struct Detail: Decodable {
            let name: String
            let url: URL
        }

        let detail: Detail
    }

    struct Move: Decodable {
        struct Detail: Decodable {
            let name: String
            let url: URL
        }

        let detail: Detail
    }

    let name: String
    let sprites: Sprites
    let types: [PokemonType]
    let moves: [Move]
}

struct PokemonDetail {
    let name: String
    let spriteURL: URL
    let types: [String]
    let attacks: [String]

    init(response: PokemonDetailResponse) {
        self.name = response.name
        self.spriteURL = response.sprites.frontDefault
        self.types = response.types.map { $0.detail.name }
        self.attacks = response.types.map { $0.detail.name }
    }
}
```